### ISSUE TITLE
Something weird with type inference in promises

### ISSUE THREAD

**Issue by forki** [fable-compiler/fable-powerpack#18]
the following code doesn't compile (which is expected because I messed up types):
```fs
    type ImageType =
    | NoImage
    | Image of string

    let takePicture() = async { return Some "" }
    let requestImageForConfirmation () = async {
        match 1 with
        | 2 ->
            return ImageType.NoImage    
        | _ -> return! takePicture()
    }
```
but the same version with promises does (but should not):
```fs
    type ImageType =
    | NoImage
    | Image of string

    let takePicture() = promise { return Some "" }
    let requestImageForConfirmation () = promise {
        match 1 with
        | 2 ->
            return ImageType.NoImage    
        | _ -> return! takePicture()  // should fail here
    }
```

**Comment by forki** [comment]
//cc @dsyme this is a bug in lib and not in compiler right?

**Comment by dsyme** [comment]
@forki Sure to be.

**Comment by forki** [comment]
any ideas where it might go wrong in https://github.com/fable-compiler/fable-powerpack/blob/master/src/Promise.fs ?
thx

**Comment by dsyme** [comment]
@forki  THe return types of ``Return`` and ``ReturnFrom`` look completely unconstrained https://github.com/fable-compiler/fable-powerpack/blob/master/src/Promise.fs#L85

**Comment by forki** [comment]
@alfonsogarciacaro can you please try to fix it?

**Comment by alfonsogarciacaro** [comment]
Done! Published as 1.1.1. Thanks a lot @dsyme for the solution!
